frontend/src/metabase/query_builder/components/expressions/ExpressionEditorTextfield.jsx

```diff
@@ -253,25 +253,29 @@ export default class ExpressionEditorTextfield extends React.Component {
 
     this.clearSuggestions();
 
-    const { query, startRule } = this.props;
-    const { source } = this.state;
-
-    const errorMessage = diagnose(source, startRule, query);
-    this.setState({ errorMessage });
-
-    // whenever our input blurs we push the updated expression to our parent if valid
-    if (errorMessage) {
-      this.props.onError(errorMessage);
-    } else {
+    try {
+      // whenever our input blurs we push the updated expression to our parent if valid
       const expression = this.compileExpression();
       if (expression) {
         if (!isExpression(expression)) {
           console.warn("isExpression=false", expression);
         }
         this.props.onChange(expression);
       } else {
-        this.props.onError({ message: t`Invalid expression` });
+        const diagnostic = this.diagnoseExpression();
```

Comment:
This seems to change the order (vs before the PR): compile first, if it failed then try to diagnose.
Please double check that `compileExpression` does not tolerate certain type of invalid expression (e.g. missing exponent digits).

Reply:
I thought we wanted to only call the new compiler only when the old one failed, just in case the newer one is more permissive?

Reply:
Reverted back to the old code, which reduced this PR's impact on `ExpressionEditorTextField.jsx`